datafusion/optimizer/src/analyzer/type_coercion.rs

```diff
@@ -99,25 +102,75 @@ fn analyze_internal(
     // select t2.c2 from t1 where t1.c1 in (select t2.c1 from t2 where t2.c2=t1.c3)
     schema.merge(external_schema);
 
-    let mut expr_rewrite = TypeCoercionRewriter { schema: &schema };
-
-    let new_expr = plan
-        .expressions()
-        .into_iter()
-        .map(|expr| {
-            // ensure aggregate names don't change:
-            // https://github.com/apache/datafusion/issues/3555
-            rewrite_preserving_name(expr, &mut expr_rewrite)
-        })
-        .collect::<Result<Vec<_>>>()?;
-
-    plan.with_new_exprs(new_expr, new_inputs)
+    let mut expr_rewrite = TypeCoercionRewriter::new(&schema);
+
+    let name_preserver = NamePreserver::new(&plan);
+    // apply coercion rewrite all expressions in the plan indivdually
+    plan.map_expressions(|expr| {
+        let original_name = name_preserver.save(&expr)?;
+        expr.rewrite(&mut expr_rewrite)?
+            .map_data(|expr| original_name.restore(expr))
+    })?
+    // coerce join expressions specially
+    .map_data(|plan| expr_rewrite.coerce_joins(plan))?
+    // recompute the schema after the expressions have been rewritten as the types may have changed
+    .map_data(|plan| plan.recompute_schema())
```

Comment:
I'm wondering if we always need to run `plan.recompute_schema()`? If the `Transformed<LogicalPlan>`'s `.transformed` is false then probably we don't need to.

Reply:
This is an excellent point. At the moment, I think we do need to always run recompute_schema because the `TypeCoercionRewriter` doesn't return `Transformed` (and thus we don't know if any actual expression coercion was done, so we have to assume it was).

I filed https://github.com/apache/datafusion/issues/10365 to track this

Reply:
Hmm, I think you use `TypeCoercionRewriter` in `expr.rewrite(&mut expr_rewrite)?` and that `rewrite()` returns `Transformed<Expr>` and then that `Transformed<Expr>` is propagated up into `plan.map_expressions()`, that returns `Transformed<LogicalPlan>`. So you have the necessary `Transformed` to decide if `recompute_schema()` is needed.

Reply:
You are correct (of course!) thank you for pointing it out.  Now that `analyze_internal` returns Transformed would work. However,  there is still code like this:

```rust
                let new_plan =
                    analyze_internal(self.schema, unwrap_arc(subquery.subquery))?.data;
                Ok(Transformed::yes(Expr::Exists(Exists {
                    subquery: Subquery {
                        subquery: Arc::new(new_plan),
                        outer_ref_columns: subquery.outer_ref_columns,
                    },
                    negated,
                })))
            }
```

Which discards the transformed information (and in this case always returns Transformed::true).


In order to keep the PRs small and easier to review I would like to not change this PR (it is no worse than `main` in regards to recomputing schema) and I will make a follow on PR to avoid recomputing schema when unecessary

Reply:
Ah ok, it seems there are many unnecessary `Transformed::yes`s in the current code. But false positive `transformed`s doesn't cause any issue...

Sure, a follow-up PR sounds good, I agree that this PR already looks really nice!

Reply:
Here is my draft followup: https://github.com/apache/datafusion/pull/10369 

It is quite large (it requires updating the entire expression rewriter) so I am glad we left it in a separate PR